Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -14,21 +14,179 @@
+
+        // Poll Kafka every millisecond
+        self.pollTask = Task { [kafkaHandle] in
+            while !Task.isCancelled {
+                rd_kafka_poll(kafkaHandle, 0)
+                try? await Task.sleep(nanoseconds: 1_000_000)
+            }
+        }
+    }
+
+    /// Kafka producer must be closed **explicitly**.
+    public func shutdownGracefully() {
+        // Wait 10 seconds for outstanding messages to be sent and callbacks to be called
+        rd_kafka_flush(self.kafkaHandle, 10000)
+
+        for (_, topicHandle) in self.topicHandles {
+            rd_kafka_topic_destroy(topicHandle)
+        }
+        self.pollTask.cancel()
+
+        _ = Unmanaged<KafkaProducer>.fromOpaque(self.referenceToSelf).takeRetainedValue()
+    }
+
+    /// Send messages to the Kafka cluster asynchronously, aka "fire and forget".
+    /// This function is non-blocking.
+    /// - Parameter message: The ``KafkaProducerMessage`` that is sent to the KafkaCluster.
+    public func sendAsync(message: KafkaProducerMessage) throws {
+        let topicHandle = self.createTopicHandleIfNeeded(topic: message.topic)
+
+        let keyBytes: [UInt8]?
+        if let key = message.key {
+            keyBytes = key.withUnsafeBytes { Array($0) }
+        } else {
+            keyBytes = nil
+        }
+
+        self.messageIDCounter += 1
+
+        let responseCode = message.value.withUnsafeBytes { valueBuffer in
+
+            return rd_kafka_produce(
+                topicHandle,
+                message.partition.rawValue,
+                RD_KAFKA_MSG_F_COPY,
+                UnsafeMutableRawPointer(mutating: valueBuffer.baseAddress),
+                valueBuffer.count,
+                keyBytes,
+                keyBytes?.count ?? 0,
+                UnsafeMutableRawPointer(bitPattern: messageIDCounter)
+            )
+        }
+
+        guard responseCode == 0 else {
+            throw KafkaError(rawValue: responseCode)
+        }
+    }
+
+    // Closure that is executed when a message has been acknowledged by Kafka
+    private let deliveryReportCallback: (
+        @convention(c) (OpaquePointer?, UnsafePointer<rd_kafka_message_t>?, UnsafeMutableRawPointer?) -> Void
+    ) = { _, messagePointer, producerPointer in
+
+        guard let messagePointer = messagePointer else {
+            return
+        }
+
+        guard let producerPointer = producerPointer else {
+            return
+        }
+        let producer = Unmanaged<KafkaProducer>.fromOpaque(producerPointer).takeUnretainedValue()
+
+        let messageID = messagePointer.pointee._private
+        guard let message = KafkaAckedMessage(messagePointer: messagePointer, id: UInt(bitPattern: messageID)) else {
+            Task {
+                producer.logger.error("Sending message failed due to an unknown error")
+            }
+            return
+        }
+
+        // TODO: is this thread-safe???
+        producer.acknowledgements.produceMessage(message)
```

Comment:
Do you think this is safe to do?

Reply:
Yes this is thread safe because the underlying `NIOAsyncSequenceProducer` is thread safe.